pkg/pipelineascode/cancel_pipelineruns.go

```diff
@@ -137,22 +144,40 @@ func (p *PacRun) cancelPipelineRuns(ctx context.Context, repo *v1alpha1.Reposito
 		return nil
 	}
 
-	var wg sync.WaitGroup
-	for _, pr := range prs.Items {
+	p.cancelPipelineRuns(ctx, prs, repo, func(pr tektonv1.PipelineRun) bool {
 		if p.event.TargetCancelPipelineRun != "" {
 			if prName, ok := pr.GetAnnotations()[keys.OriginalPRName]; !ok || prName != p.event.TargetCancelPipelineRun {
-				continue
+				return false
 			}
 		}
-		if pr.IsDone() {
-			p.logger.Infof("pipelinerun %v/%v is done, skipping cancellation", pr.GetNamespace(), pr.GetName())
+		return true
+	})
+
+	return nil
+}
+
+func (p *PacRun) cancelPipelineRuns(ctx context.Context, prs *tektonv1.PipelineRunList, repo *v1alpha1.Repository, condition matchingCond) {
+	var wg sync.WaitGroup
+	for _, pr := range prs.Items {
+		if !condition(pr) {
 			continue
 		}
+
 		if pr.IsCancelled() || pr.IsGracefullyCancelled() || pr.IsGracefullyStopped() {
-			p.logger.Infof("pipelinerun %v/%v is already in %v state", pr.GetNamespace(), pr.GetName(), pr.Spec.Status)
+			p.logger.Infof("cancel-in-progress: skipping cancelling pipelinerun %v/%v, already in %v state", pr.GetNamespace(), pr.GetName(), pr.Spec.Status)
 			continue
 		}
 
+		if pr.IsDone() {
+			p.logger.Infof("cancel-in-progress: skipping cancelling pipelinerun %v/%v, already done", pr.GetNamespace(), pr.GetName())
+			continue
+		}
+
+		if pr.IsPending() {
```

Comment:
Any reason to skip pending ones ? Would it match pipelineruns from new commits/push ?

Reply:
if it's pending there is nothing to cancel? 

deleting the prun directly could be an issue for the concurrency q and i'd like to don't mess with that...

Reply:
and furthermore cancel-in-progress is explicitely documented as not working with the concurrency feature (mostly because of the complexitiy involved for such a simple implementation) so that should not be supported...

Reply:
@chmouel but a pending pipelinerun is waiting to be started *or* cancelled. If we do not cancel, it could stay here forever.